### PR TITLE
✨(CourseRunEnrollment) customized failure message

### PR DESCRIPTION
**Pull request by sandroscosta**
Related to #1871 

## Proposal
Give Richie the possibility of displaying custom error messages that come from an error payload

### CODE REVIEW

**Comment by sandroscosta** [comment]
@jbpenrath There seems to be an issue with `error` getting to the correct place. Apparently error is always undefined. Don't know if we need to change the API adapter to accept an error with a payload. Any hints on where to go next or if something is wrong.

File: src/frontend/js/components/CourseRunEnrollment/index.tsx [hunk 1]
```diff
@@ -220,7 +230,7 @@ const CourseRunEnrollment: React.FC<CourseRunEnrollmentProps & CommonDataProps>
           </button>
           {step === Step.ENROLLMENT_FAILED ? (
             <div className="course-run-enrollment__errortext">
-              <FormattedMessage {...messages.enrollmentFailed} />
+              {error != undefined && error.code == 400 ? <>{error.message}</> : <FormattedMessage {...messages.enrollmentFailed} />}
```
Comment (jbpenrath) [inline src/frontend/js/components/CourseRunEnrollment/index.tsx:233]: Use strict operators. Furthermore, you could use an optional chaining operator (`?`) instead to test if `error` is defined.

```suggestion
              {error?.code === 400 ? error.message : <FormattedMessage {...messages.enrollmentFailed} />}

```

File: src/frontend/js/components/CourseRunEnrollment/index.tsx [hunk 2]
```diff
@@ -83,12 +85,14 @@ interface ReducerState {
     isEnrolled: Maybe<Nullable<Boolean>>;
     courseRun: CourseRunEnrollmentProps['courseRun'];
     currentUser: Maybe<Nullable<User>>;
+    error?: HttpError;
```
Comment (jbpenrath) [inline src/frontend/js/components/CourseRunEnrollment/index.tsx:88]: There is already an `error` property outside context.

File: src/frontend/js/components/CourseRunEnrollment/index.tsx [hunk 3]
```diff
@@ -132,6 +136,8 @@ const reducer = ({ step, context }: ReducerState, action: ReducerAction): Reduce
       const nextContext = { ...context, ...action.payload };
       return { step: getStepFromContext(nextContext, step), context: nextContext };
     }
+    case ActionType.ENROLLMENT_FAILED:
+      return { step: Step.ENROLLING, context: {...context, ...action.payload }};
```
Comment (jbpenrath) [inline src/frontend/js/components/CourseRunEnrollment/index.tsx:140]:
```suggestion
      return { step: Step.ENROLLMENT_FAILED, context, ...action.payload };
```

File: src/frontend/js/components/CourseRunEnrollment/index.tsx [hunk 4]
```diff
@@ -161,7 +166,13 @@ const CourseRunEnrollment: React.FC<CourseRunEnrollmentProps & CommonDataProps>
   const enroll = useCallback(async () => {
     dispatch({ type: ActionType.ENROLL });
     if (courseRun && currentUser) {
-      const isEnrolled = await setEnrollment().catch(() => undefined);
+      const isEnrolled = await setEnrollment().catch((error) => {
+        dispatch({
+          type: ActionType.ENROLLMENT_FAILED,
+          payload: { error },
+        });
+        return undefined;
```
Comment (jbpenrath) [inline src/frontend/js/components/CourseRunEnrollment/index.tsx:174]: Return statement is useless.

```suggestion
```

File: src/frontend/js/components/CourseRunEnrollment/index.tsx [hunk 5]
```diff
@@ -161,7 +166,13 @@ const CourseRunEnrollment: React.FC<CourseRunEnrollmentProps & CommonDataProps>
   const enroll = useCallback(async () => {
     dispatch({ type: ActionType.ENROLL });
     if (courseRun && currentUser) {
-      const isEnrolled = await setEnrollment().catch(() => undefined);
+      const isEnrolled = await setEnrollment().catch((error) => {
+        dispatch({
```
Comment (jbpenrath) [inline src/frontend/js/components/CourseRunEnrollment/index.tsx:170]: I think you should use a `try...catch` here.
Furthermore there is something wrong as the action `UPDATE_CONTEXT` is always called. You should dispatch this action only when enroll request succeeded.

**Comment by sandroscosta** [comment]
Had to change all Errors to be `HttpError` to be able to access code.

**Comment by sandroscosta** [comment]
Deprecated to switch to a new approach

**Pull request closed without merging**